File: projects/finance-copilot/src/finance_copilot/llm/gigachat.py
"""Провайдер GigaChat (Sber).

Док: https://developers.sber.ru/docs/ru/gigachat/api/overview
Аутентификация: по Authorization Key получаем короткоживущий access_token,
затем ходим в chat/completions. Токен кэшируем в памяти до истечения.
"""

from __future__ import annotations

import logging
import time
import uuid

import requests

from .base import LLMProvider

logger = logging.getLogger(__name__)

_OAUTH_URL = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth"
_CHAT_URL = "https://gigachat.devices.sberbank.ru/api/v1/chat/completions"


class GigaChatProvider(LLMProvider):
    name = "gigachat"

    def __init__(
        self,
        auth_key: str,
        scope: str = "GIGACHAT_API_PERS",
        model: str = "GigaChat",
        timeout: int = 30,
        verify_ssl: bool = True,
    ) -> None:
        if not auth_key:
            raise ValueError("Для GigaChat нужен GIGACHAT_AUTH_KEY")
        self.auth_key = auth_key
        self.scope = scope
        self.model = model
        self.timeout = timeout
        # В корпоративных сетях иногда требуется цепочка сертификатов Минцифры;
        # verify_ssl оставлен настраиваемым, но по умолчанию проверка включена.
        self.verify_ssl = verify_ssl
        self._token: str | None = None
        self._token_exp: float = 0.0

    def _ensure_token(self) -> str:
        if self._token and time.monotonic() < self._token_exp:
            return self._token
        headers = {
            "Authorization": f"Basic {self.auth_key}",
            "RqUID": str(uuid.uuid4()),
            "Content-Type": "application/x-www-form-urlencoded",
        }
        resp = requests.post(
            _OAUTH_URL,
            headers=headers,
            data={"scope": self.scope},
            timeout=self.timeout,
            verify=self.verify_ssl,
        )
        resp.raise_for_status()
        data = resp.json()
        self._token = data["access_token"]
        # expires_at приходит в миллисекундах epoch; берём запас 30 сек.
        self._token_exp = time.monotonic() + 1500
        return self._token

    def complete(self, prompt: str, system: str | None = None) -> str:
        token = self._ensure_token()
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})
        payload = {"model": self.model, "messages": messages, "temperature": 0.2}
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        }
        resp = requests.post(
            _CHAT_URL,
            json=payload,
            headers=headers,
            timeout=self.timeout,
            verify=self.verify_ssl,
        )
        resp.raise_for_status()
        data = resp.json()
        return data["choices"][0]["message"]["content"]

File: projects/marketplace-copilot/lib/llm/gigachat.ts
import { randomUUID } from "crypto";
import { ChatMessage, CompletionOptions, LLMError, LLMProvider } from "./types";

const OAUTH_URL = "https://ngw.devices.sberbank.ru:9443/api/v2/oauth";
const CHAT_URL =
  "https://gigachat.devices.sberbank.ru/api/v1/chat/completions";

interface CachedToken {
  accessToken: string;
  /** Unix ms, когда токен истекает. */
  expiresAt: number;
}

/**
 * GigaChat (Сбер) — основной провайдер.
 *
 * Схема авторизации двухступенчатая:
 *  1. По Authorization key (base64 client_id:client_secret) получаем короткоживущий
 *     OAuth access-токен на выбранный scope.
 *  2. Access-токен используем как Bearer для чат-запросов.
 *
 * Токен кэшируется в памяти процесса и переиспользуется до истечения.
 *
 * Прим.: у GigaChat самоподписанный корневой сертификат НУЦ Минцифры. В проде
 * его добавляют в доверенные (NODE_EXTRA_CA_CERTS). Для локальной разработки
 * это описано в README; тут мы не отключаем проверку TLS в коде.
 */
export class GigaChatProvider implements LLMProvider {
  readonly name = "gigachat";
  private token: CachedToken | null = null;

  constructor(
    private readonly authKey: string,
    private readonly scope: string,
    private readonly model: string,
  ) {
    if (!authKey) {
      throw new LLMError(
        "GIGACHAT_AUTH_KEY не задан в окружении",
        "gigachat",
      );
    }
  }

  private async getAccessToken(): Promise<string> {
    const now = Date.now();
    // Обновляем заранее — за 60 сек до истечения.
    if (this.token && this.token.expiresAt - 60_000 > now) {
      return this.token.accessToken;
    }

    const body = new URLSearchParams({ scope: this.scope });
    let res: Response;
    try {
      res = await fetch(OAUTH_URL, {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          Accept: "application/json",
          RqUID: randomUUID(),
          Authorization: `Basic ${this.authKey}`,
        },
        body,
      });
    } catch (e) {
      throw new LLMError(
        "Не удалось подключиться к OAuth GigaChat (проверьте сеть и TLS-сертификат НУЦ Минцифры)",
        "gigachat",
        e,
      );
    }

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      throw new LLMError(
        `OAuth GigaChat вернул ${res.status}: ${text}`,
        "gigachat",
      );
    }

    const data = (await res.json()) as {
      access_token: string;
      expires_at: number;
    };
    this.token = {
      accessToken: data.access_token,
      // expires_at приходит в Unix ms.
      expiresAt: data.expires_at,
    };
    return this.token.accessToken;
  }

  async complete(
    messages: ChatMessage[],
    options: CompletionOptions = {},
  ): Promise<string> {
    const accessToken = await this.getAccessToken();

    let res: Response;
    try {
      res = await fetch(CHAT_URL, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
          Authorization: `Bearer ${accessToken}`,
        },
        body: JSON.stringify({
          model: this.model,
          messages,
          temperature: options.temperature ?? 0.3,
          max_tokens: options.maxTokens ?? 1500,
        }),
      });
    } catch (e) {
      throw new LLMError("Сетевая ошибка запроса к GigaChat", "gigachat", e);
    }

    if (!res.ok) {
      const text = await res.text().catch(() => "");
      throw new LLMError(
        `GigaChat вернул ${res.status}: ${text}`,
        "gigachat",
      );
    }

    const data = (await res.json()) as {
      choices?: { message?: { content?: string } }[];
    };
    const content = data.choices?.[0]?.message?.content;
    if (!content) {
      throw new LLMError("GigaChat вернул пустой ответ", "gigachat");
    }
    return content.trim();
  }
}
